<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>1&nbsp; The use-case data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/02_import_and_infrastructure.html" rel="next">
<link href="../setup.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Analysis of expression proteomics data in R</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/CambridgeCentreForProteomics/course_expression_proteomics" rel="" target=""><i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs" rel="" target=""><i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../materials/01_intro.html">Expression proteomics</a></li><li class="breadcrumb-item"><a href="../materials/01_intro.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The use-case data</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Welcome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data &amp; Setup</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Expression proteomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/01_intro.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The use-case data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/02_import_and_infrastructure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Import and infrastructure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03_data_processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data cleaning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04_normalisation_aggregation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data normalisation and data aggregation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/05_protein_exploration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Exploration and visualisation of protein data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/06_statistical_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Statistical analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/07_biological_interpretation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Biological interpretation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Extended materials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/08_maxquant_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Adapting this workflow to MaxQuant processed data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/09_normalyzer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Exploring normalisation methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/10_lfq_vs_tmt.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Adapting this workflow to label-free proteomics data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/11_statistical_analysis_all_stages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Statistical analysis of all cell cycle stages</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#exploring-changes-in-protein-abundance-across-stages-of-the-cell-cycle" id="toc-exploring-changes-in-protein-abundance-across-stages-of-the-cell-cycle" class="nav-link active" data-scroll-target="#exploring-changes-in-protein-abundance-across-stages-of-the-cell-cycle"><span class="header-section-number">1.1</span> Exploring changes in protein abundance across stages of the cell cycle</a></li>
  <li><a href="#isobaric-peptide-labelling-using-tandem-mass-tags-tmt" id="toc-isobaric-peptide-labelling-using-tandem-mass-tags-tmt" class="nav-link" data-scroll-target="#isobaric-peptide-labelling-using-tandem-mass-tags-tmt"><span class="header-section-number">1.2</span> Isobaric peptide labelling using Tandem Mass Tags (TMT)</a></li>
  <li><a href="#identification-search-of-the-raw-ms-data" id="toc-identification-search-of-the-raw-ms-data" class="nav-link" data-scroll-target="#identification-search-of-the-raw-ms-data"><span class="header-section-number">1.3</span> Identification search of the raw MS data</a></li>
  <li><a href="#starting-analysis-from-psm-peptide-or-protein-level" id="toc-starting-analysis-from-psm-peptide-or-protein-level" class="nav-link" data-scroll-target="#starting-analysis-from-psm-peptide-or-protein-level"><span class="header-section-number">1.4</span> Starting analysis from PSM, peptide or protein level</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">1.5</span> Summary</a></li>
  <li><a href="#glossary" id="toc-glossary" class="nav-link" data-scroll-target="#glossary"><span class="header-section-number">1.6</span> Glossary</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The use-case data</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Understand the aim and design of a common expression proteomics experiment</li>
<li>Know how relative quantitation can be achieved using Tandem Mass Tag (TMT) labelling</li>
<li>Know what a database/identification search is and what data level to output and begin processing your data</li>
</ul>
</div>
</div>
<section id="exploring-changes-in-protein-abundance-across-stages-of-the-cell-cycle" class="level2" data-number="1.1"><h2 data-number="1.1" class="anchored" data-anchor-id="exploring-changes-in-protein-abundance-across-stages-of-the-cell-cycle">
<span class="header-section-number">1.1</span> Exploring changes in protein abundance across stages of the cell cycle</h2>
<p>As a use case, we will analyse an expression proteomics dataset from <span class="citation" data-cites="Queiroz2019">Queiroz et al. (<a href="#ref-Queiroz2019" role="doc-biblioref">2019</a>)</span>. The aim of this experiment was to examine proteins involved in the cell cycle (see <a href="#fig-experimental-method">Figure&nbsp;<span>1.1</span></a>). Briefly, cells from a human cell line (U-2 0S) were treated with Nocodazole to inhibit the polymerisation of microtubules and block the transition beyond the prometaphase stage of M-phase. After 18-hours of treatment, the cells were released from cell cycle inhibition by withdrawing Nocodazole. Samples were taken in triplicate at 0, 6 and 23-hours post-withdrawl to generate 3 x M-phase samples, 3 x G1-phase samples and 3 x desynchronised samples. A sample of cells was also taken prior to Nocodazole treatment as an additional control, thus giving a total of 10 samples.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-experimental-method" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figs/experimental_plan_updated.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.1: A schematic summary of the experimental design used to generate the use case data.</figcaption></figure>
</div>
</div>
</div>
<p>To measure protein abundance via bottom-up mass spectrometry, protein samples were digested to peptides using trypsin (an enzyme which cleaves proteins at the C-terminus of arginine and lysine residues, except where there is an adjacent proline residue). The same quantity of peptide from each sample was then labelled using Tandem Mass Tag (TMT) labels. This allows all of the samples to be combined into a single pooled MS sample, thus reducing technical variability and MS time, as well as ensuring quantification of the same peptides in each samples. The TMT labelling strategy is outlined below in <a href="#tbl-data-structure">Table&nbsp;<span>1.1</span></a>.</p>
<div class="cell">
<div class="cell-output-display">
<div id="tbl-data-structure" class="anchored">
<table class="table table-sm table-striped small">
<caption>Table&nbsp;1.1: Sample information and TMT labelling strategy in the use case experiment</caption>
<thead><tr class="header">
<th style="text-align: left;">Sample Name</th>
<th style="text-align: left;">Timepoint</th>
<th style="text-align: right;">Replicate</th>
<th style="text-align: left;">Tag</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Control</td>
<td style="text-align: left;">Pre-treatment</td>
<td style="text-align: right;">NA</td>
<td style="text-align: left;">126</td>
</tr>
<tr class="even">
<td style="text-align: left;">M_1</td>
<td style="text-align: left;">0hrs</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">127N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">M_2</td>
<td style="text-align: left;">0hrs</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">127C</td>
</tr>
<tr class="even">
<td style="text-align: left;">M_3</td>
<td style="text-align: left;">0hrs</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">128N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">G1_1</td>
<td style="text-align: left;">6hrs</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">128C</td>
</tr>
<tr class="even">
<td style="text-align: left;">G1_2</td>
<td style="text-align: left;">6hrs</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">129N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">G1_3</td>
<td style="text-align: left;">6hrs</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">129C</td>
</tr>
<tr class="even">
<td style="text-align: left;">DS_1</td>
<td style="text-align: left;">23hrs</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">130N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DS_2</td>
<td style="text-align: left;">23hrs</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">130C</td>
</tr>
<tr class="even">
<td style="text-align: left;">DS_3</td>
<td style="text-align: left;">23hrs</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">131</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section><section id="isobaric-peptide-labelling-using-tandem-mass-tags-tmt" class="level2" data-number="1.2"><h2 data-number="1.2" class="anchored" data-anchor-id="isobaric-peptide-labelling-using-tandem-mass-tags-tmt">
<span class="header-section-number">1.2</span> Isobaric peptide labelling using Tandem Mass Tags (TMT)</h2>
<p>As outlined above, the use-case experiment made use of tandem mass tag (TMT) labelling. TMT reagents are <strong>isobaric</strong> chemical labels that can be covalently linked to peptide amine groups, either at the peptide N-terminus or at lysine (K) residues. These reagents are currently available in kits (referred to as TMTplexes) containing 6 (TMT6plex<sup>TM</sup>), 10 (TMT10plex<sup>TM</sup>), 16 (TMTpro 16plex<sup>TM</sup>), 18 (TMTpro 18plex<sup>TM</sup>) or 32 (TMTpro 32plex<sup>TM</sup>) TMT labels. Further, kits can be combined to multiplex up to 35 samples. The isobaric nature of TMT reagents means that all of the labels within a TMTplex have the same overall mass. However, they differ in how this mass is distributed between their two variable regions - the <strong>reporter</strong> and <strong>mass normaliser</strong>. The structure of a TMT label is shown below in <a href="#fig-tmt-structure">Figure&nbsp;<span>1.2</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-tmt-structure" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figs/tmt_structure.jpg" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.2: The chemical structure of a TMT label.</figcaption></figure>
</div>
</div>
</div>
<p>After labelling each sample with a different TMT reagent, the same peptide will have an identical mass but be differentially labelled across samples. The samples are then pooled together and run on the MS as a single <strong>multiplexed</strong> sample. In the MS1 spectra it is not possible to distinguish how much of each peptide (peak) was derived from each sample because all of the labelled peptides still have the same overall mass. During fragmentation, however, a cleavable linker within the TMT label is broken, thereby releasing a section of the label called the <strong>reporter ion</strong>. The mass of this reporter ion is different between TMT labels due to differences in the distribution of isotopes (C<sup>13</sup> and N<sup>15</sup>) between the <strong>mass reporter</strong> and <strong>mass normalizer</strong> regions. As a result, the spectra of fragment ions will contain one peak per reporter ion and the relative intensity of these ions can be used to calculate the relative abundance of the identified peptide across the labelled samples. This is summarised below in <a href="#fig-tmt-quant">Figure&nbsp;<span>1.3</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-tmt-quant" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><img src="figs/tmt_quant.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.3: Relative peptide quantitation using TMT labels. Image modified from Thermo Fisher Scientific.</figcaption></figure>
</div>
</div>
</div>
<p>This means that TMT labelling can be used for the <strong>relative quantification</strong> of peptides (and in turn proteins) across samples. Here we will use the relative reporter ion intensity of our TMT labels to compare abundance between samples taken throughout the cell cycle. If we wanted to carry out absolute quantification of proteins, TMT labelling would not be an appropriate method.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
MS2- vs.&nbsp;MS3-based TMT quantitaion
</div>
</div>
<div class="callout-body-container callout-body">
<p>Originally the reporter ion of a TMT reagent was released by fragmentation of the precursor peptide. This meant that the reporter ion signals could be calculated based on MS2 spectra, and hence TMT quantitation was done at the MS2 level. However, different precursor peptides can have very similar m/z values, resulting in their MS1 peaks being very close together. When one of these precursors is selected for fragmentation it is possible for other precursors to be unintentionally co-isolated within the same fragmentation window. This results in <strong>co-isolation interference</strong> since the reporter ions present at MS2 are not only from the identified peptide but also any other co-isolated peptides.</p>
<p>More recently, MS3-based quantitation has been applied to reduce the problem of co-isolation interference and increase quantitation accuracy when using TMT labels <span class="citation" data-cites="Ting2011">Ting et al. (<a href="#ref-Ting2011" role="doc-biblioref">2011</a>)</span>. In this approach the initial round of precursor fragmentation only uses a weak fragmentation method (e.g., collision induced dissociation, CID) to ensure that TMT reporter ions are not broken off at MS2. The most abundance fragment ions from MS2 are then selected for a second round of fragmentation, this time using a stronger force (e.g., high-energy collision dissociation) so that the TMT reporter ions are released. Since the MS2 fragment ions derived from the intended precursor peptide are more abundant than the unwanted MS2 fragments from co-isolated peptides, the MS3 fragments and reporter ions provide a much cleaner quantitation of the identified precursor peptide. The MS3 reporter ion quantitation is almost purely from the peptide that we want, not from a mixture of all co-isolated peptides.</p>
<p>The specific MS3 data acquisition approach used to generate our data was synchronous precursor selection (SPS)-MS3. In simple terms, the SPS method allows us to select multiple MS2 fragments for additional fragmentation to the MS3 level, rather than just one <span class="citation" data-cites="McAlister2014">McAlister et al. (<a href="#ref-McAlister2014" role="doc-biblioref">2014</a>)</span>.</p>
</div>
</div>
</section><section id="identification-search-of-the-raw-ms-data" class="level2" data-number="1.3"><h2 data-number="1.3" class="anchored" data-anchor-id="identification-search-of-the-raw-ms-data">
<span class="header-section-number">1.3</span> Identification search of the raw MS data</h2>
<p>Having analysed the pooled TMT sample via MS, the raw MS were processed using Proteome Discoverer v3.0 (Thermo Fisher Scientific). This software is one of several that can be used to carry out a database search to identify and quantify peptide sequences, and therefore proteins, from raw MS data. Others include <a href="https://www.maxquant.org">MaxQuant</a>, <a href="https://fragpipe.nesvilab.org">FragPipe</a> and <a href="https://www.bioinfor.com/peaks-online/">PEAKS</a> among others.</p>
<p>Briefly, database searching of MS data involves comparing the observed MS spectra to expected MS spectra generated via in silico digestion of a selected protein database. Since the use-case data used human cells, we provided the human proteome to Proteome Discoverer as the database for searching. We also provided a database of common contaminants that could be present in our sample due to (i) human contamination (e.g., keratin from nails, hair or skin), or (ii) sample preparation (e.g., enzymes used for protein digestion). These two databases are can be found in the course materials and also at <a href="https://zenodo.org/records/7837375">https://zenodo.org/records/7837375</a>. The result of such a database search is a list of <strong>peptide spectrum matches (PSMs)</strong>, that is a list of matches between observed and expected spectra. Most software will also aggregate the PSM level data upward and provide output files at all data levels (PSM, peptide and protein). It is up to the user which output file to use for further analysis.</p>
<p>The starting point for this workshop and data processing workflow in R is an output file from the identification search. Proteome Discoverer provides each data level output as a separate <code>.txt</code> file. Here, we will use the PSM-level <code>.txt</code> file.</p>
</section><section id="starting-analysis-from-psm-peptide-or-protein-level" class="level2" data-number="1.4"><h2 data-number="1.4" class="anchored" data-anchor-id="starting-analysis-from-psm-peptide-or-protein-level">
<span class="header-section-number">1.4</span> Starting analysis from PSM, peptide or protein level</h2>
<p>In general, to allow for maximum control of data quality, normalisation and aggregation it is advisable to begin analysis from the lowest possible data level. This will ensure a greater understanding of the data and facilitate transparency throughout the process.</p>
<p>For TMT data we can start our data analysis from the PSM level. However, it is not always possible to start at this level. Some algorithms used to process label-free data may require the data to be analysed from the peptide-level. See <a href="./11_lfq_vs_tmt.html">Adapting this workflow to label-free proteomics</a> for more details.</p>
</section><section id="summary" class="level2" data-number="1.5"><h2 data-number="1.5" class="anchored" data-anchor-id="summary">
<span class="header-section-number">1.5</span> Summary</h2>
<p>The use-case data that we will process and analyse in this workshop is a <strong>DDA</strong> <strong>TMT-labelled</strong> <strong>bottom-up</strong> proteomics dataset. The aim of the experiment is to calculate <strong>relative protein abundances</strong> between samples. This will then allow us to apply statistical tests to determine whether any proteins have significantly different abundances relative to each other.</p>
<p>Although many of the processing and analysis steps discussed in this course are also applicable to other types of proteomics datasets (particularly label-free DDA bottom-up proteomics), there is no one-size-fits-all workflow for expression proteomics. Some aspects of the workflow would need to be adapted for other types of proteomics experiment. We have provided details on how this workflow could be adapted to <a href="./11_lfq_vs_tmt.html">label-free data</a> or <a href="./10_maxquant_notes.html">data processed using MaxQuant</a> rather than Proteome Discoverer.</p>
</section><section id="glossary" class="level2" data-number="1.6"><h2 data-number="1.6" class="anchored" data-anchor-id="glossary">
<span class="header-section-number">1.6</span> Glossary</h2>
<p>Key terms that you should understand for this workshop:</p>
<ul>
<li><p><strong>Precursor ion</strong> = the original parent ion representing an ionized form of the entire peptide sequence.</p></li>
<li><p><strong>Fragment ion</strong> = an ion produced by fragmentation of the precursor ion, thus only representing a fraction of the original peptide sequence.</p></li>
<li><p><strong>MS1 spectrum</strong> = raw mass spectrum produced by the separation of precursor ions based on their mass-to-charge ratio (m/z). Each peak represents a precursor ion at a particular m/z and with an associated intensity.</p></li>
<li><p><strong>MS2 (MS/MS) spectrum</strong> = raw mass spectrum produced by the separation of fragment ions based on their mass-to-charge ratio (m/z). Each peak corresponds to a fragment ion derived from the same precursor ion.</p></li>
<li><p><strong>Peptide spectrum match (PSM)</strong> = A match made between a theoretical mass spectrum for a given peptide sequence and an observed experimental spectrum, thus linking a raw mass spectrum to its predicted peptide sequence</p></li>
<li><p><strong>Tandem mass tag (TMT)</strong> = a type of peptide label which can be used for relative quantification of peptides across samples. Quantification is measured at the MS2 or MS3 level.</p></li>
</ul></section><section id="references" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="references">References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-McAlister2014" class="csl-entry" role="listitem">
McAlister, Graeme C., David P. Nusinow, Mark P. Jedrychowski, Martin Wühr, Edward L. Huttlin, Brian K. Erickson, Ramin Rad, Wilhelm Haas, and Steven P. Gygi. 2014. <span>“<span>MultiNotch</span> <span>MS</span>3 Enables Accurate, Sensitive, and Multiplexed Detection of Differential Expression Across Cancer Cell Line Proteomes.”</span> <em>Analytical Chemistry</em> 86 (14): 7150–58. <a href="https://doi.org/10.1021/ac502040v">https://doi.org/10.1021/ac502040v</a>.
</div>
<div id="ref-Queiroz2019" class="csl-entry" role="listitem">
Queiroz, Rayner M. L., Tom Smith, Eneko Villanueva, Maria Marti-Solano, Mie Monti, Mariavittoria Pizzinga, Dan-Mircea Mirea, et al. 2019. <span>“Comprehensive Identification of <span>RNA</span><span></span>protein Interactions in Any Organism Using Orthogonal Organic Phase Separation (<span>OOPS</span>).”</span> <em>Nature Biotechnology</em> 37 (2): 169–78. <a href="https://doi.org/10.1038/s41587-018-0001-2">https://doi.org/10.1038/s41587-018-0001-2</a>.
</div>
<div id="ref-Ting2011" class="csl-entry" role="listitem">
Ting, Lily, Ramin Rad, Steven P Gygi, and Wilhelm Haas. 2011. <span>“<span>MS</span>3 Eliminates Ratio Distortion in Isobaric Multiplexed Quantitative Proteomics.”</span> <em>Nature Methods</em> 8 (11): 937–40. <a href="https://doi.org/10.1038/nmeth.1714">https://doi.org/10.1038/nmeth.1714</a>.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../setup.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Data &amp; Setup</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/02_import_and_infrastructure.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Import and infrastructure</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>