---
title: Statistical analysis
---

::: {.callout-tip}
#### Learning Objectives

* Acknowledge the availability of different R/Bioconductor packages for carrying out differential expression (abundance) analyses
* Using the `Limma` package, design a statistical model to test for differentially abundant proteins between two conditions
* Be able to interpret the output of a statistical model and annotate the results with user-defined significance thresholds
* Produce volcano plots and MA plots to visualise the results of differential expression analyses
* Export key results from R using `write_csv` and figures using `ggsave`

:::

```{r, eval=TRUE, include=FALSE}
library("QFeatures")
library("ggplot2")
library("stringr")
library("dplyr")
library("tibble")
library("NormalyzerDE")
library("corrplot")
library("Biostrings")
library("limma")
library("statmod")
library("org.Hs.eg.db")
library("clusterProfiler")
library("enrichplot")
load("output/lesson03.rda", verbose = TRUE)
```


## Differential expression analysis 

Having cleaned our data, aggregated from PSM to protein level, completed a
log2 transformation and normalised the data, we are now ready to carry out
statistical analysis. The aim of this section is to answer the question:
*"Which proteins show a significant change in abundance between cell cycle stages?"*.

Our null hypothesis is:
**(H0)** The change in abundance for a protein between cell cycle stages is 0.

Our alternative hypothesis is:
**(H1)** The change in abundance for a protein between cell cycle stages is greater than 0.


## Selecting a statistical test

There are a few aspects of our data that we need to consider prior to deciding
which statistical test to apply.

* The protein abundances in this data are not normally distributed (i.e., they do not follow a Gaussian distribution). However, they are approximately normal following a log2 transformation.
* The cell cycle is not expected to have a large impact on biological variability. We can assume that the majority of the proteome is the same between samples.
* The samples are independent not paired. For example, M_1 is not derived from the same cells as G1_1 and DS_1.

The first two points relate to two key assumptions that are made when carrying out 
**parametric** statistical tests. Parametric tests provide greater statistical 
power but assume that the input data has a normal distribution and equal variance.
If these two assumptions are not met then it is not appropriate to use parametric 
testing. If we wanted to check whether the log2 protein abundances were truly 
Gaussian we could have applied a Shapiro-Wilk test or Kolmogorov-Smirnov test.
Similarly, to check for equal variance a Levene test could be used. All of these
tests are easily done within R but will not be covered in this course. Here, we
will use a form of t-test since we know that these assumptions have been met.

Many different R packages can be used to carry out differential expression 
(abundance) analysis on proteomics data. Here we will use `limma`, a package that
is widely used for omics analysis and has several models that allow differential
abundance to be assessed in multifactorial experiments. Specifically, we are 
going to apply `limma`'s **empirical Bayes moderated t-test**.

A t-test is a parametric statistical test used to compare the mean values of two
groups, essentially asking whether the difference between means is significantly
greater than 0. The output of such a test includes several parameters:

* **t-value** = provides a ratio between the size of the difference between means and the variation between samples within condition. The further away from 0 that a t-value lies, the more likely it is to represent a significant difference between means (large difference between conditions, small variation within conditions)
* **degrees of freedom** = the number of observations minus the number of independent variables (*n* - 1)
* **p-value** = the probability of achieving the t-value under the null hypothesis i.e., by chance

Here, we will apply a empirical Bayes moderated t-test to each of our proteins 
independently to ask whether the difference in the mean abundance of a protein
between cell cycle stages is significantly different from 0.


### What does the empirical Bayes part mean?

When carrying out high throughput omics experiments we not only have a population
of samples but also a population of features - here we have several thousand 
proteins. 


* Explain each term - Bayes moderated, t-test
* Explain the concept of contrasts
* Discuss the two types of limma model (trend = TRUE)


## Defining the statistical model 

Before we apply our empirical Bayes moderated t-test, we first need to set up a 
model. To define the model design we use `model.matrix`. A **model matrix**, also
called a design matrix, is a matrix in which rows represent individual samples and 
columns correspond to explanatory variables, in our case the cell cycle stages.
Simply put, the model design is determined by how samples are distributed across
conditions. 


### With or without an intercept

When investigating the effect of a single explanatory variable, a design matrix
can be created using either `model.matrix(~variable)` or `model.matrix(~0 + variable)`.
The difference between these two options is that the first will create a model 
that includes an intercept term whilst the second will exclude any intercept
term. If the explantory variable is a factor, both models with and without an
intercept are equivalent. If, however, the explanatory variable is a covariate, 
the two models are then fundamentally different. 

In this experiment we consider our explanatory variable (cell cycle stage) to 
be categorical, making it of `factor` class in R. When modelling factor
explanatory variables, models with and without an intercept are equivalent. 
Either option is appropriate for factor explanatory variables, but design matrices
without an intercept value tend to be easier to interpret. For further guidance
on generating design matrices for covariates or continuous explanatory 
variables, see [A guide to creating design matrices for gene expression experiments](https://bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-and-contrast-matrices).

We subset our log2 normalised protein level data since this will be the
input for our statistical test. For simplicity we also remove the 1st sample, 
which was a single pre-treatment control. Without any replicates this condition
cannot be considered in a statistical framework.

```{r}
## Extract protein list from our QFeatures object - remove sample 1
all_proteins <- cc_qf[["log_norm_proteins"]][, -1]

## Ensure that conditions are stored as levels of a factor - check level order
all_proteins$condition <- factor(all_proteins$condition, levels = c("Desynch",
                                                                    "M",
                                                                    "G1"))

```

Next, we define the model matrix without any intercept term.

```{r}
## Design a matrix containing all factors that we wish to model
m_design <- model.matrix(~ 0 + all_proteins$condition)
colnames(m_design) <- levels(all_proteins$condition)

## Verify
m_design
```

We also define **contrasts**. The contrasts represent which comparisons or 
questions are of interest to us. This is importance since we are not directly 
interested in the parameter (mean) estimates for each group but rather the differences
in these parameter (mean) estimates between groups. The `makeContrasts` function
is used by passing the name we wish to give each contrast and how this contrast
should be calculated using column names from the design matrix. We also pass the
`levels` argument to tell R where the column names come from i.e., which design
matrix the contrasts are being applied to.


```{r}
## Specify contrasts of interest
contrasts <- makeContrasts(M_G1 = G1 - M, 
                           M_Des = Desynch - M, 
                           G1_Des = Desynch - G1,
                           levels = m_design)

## Verify
contrasts
```


## Running an empirical Bayes moderated t-test using `Limma`

After we have specified the design matrix and contrasts we wish to make, the 
next step is to apply the statistical model. 

```{r}
## Fit linear model using the design matrix and desired contrasts
fit_model <- lmFit(assay(all_proteins), m_design)
fit_contrasts <- contrasts.fit(fit_model, contrasts)

## Update the model using the limma eBayes algorithm
fit_contrasts <- eBayes(fit_contrasts, 
                        trend = TRUE,
                        robust = TRUE)
```

The initial t-test has now been applied to each of the proteins in our data. 
We now update 



```{r}
## Format results
limma_results <- topTable(fit_contrasts,
                          adjust.method = "BH",
                          number = Inf) %>%
  rownames_to_column("Protein") 

## Verify
head(limma_results)
```


### Diagnostic plots to verify suitability of our statistical model

As with all statistical analysis, it is crucial to do some quality control and
to check that the statistical test that has been applied was indeed appropriate
for the data. As mentioned above, statistical tests typically come with several
assumption. To check that these assumptions were met and that our model was
suitable, we create some diagnostic plots.


```{r}
limma_results %>%
  as_tibble() %>%
  ggplot(aes(x = P.Value)) + 
  geom_histogram()
```
```{r}
plotSA(fit_contrasts,
       cex = 0.5,
       xlab = "Average log2 abundance")
```

* p-value histogram 
* SA plot


### Interpreting the output of our statisical model

```{r}
for (i in colnames(contrasts)) {
  
  results <- topTable(fit_contrasts, coef = i, number = Inf, adjust.method = "BH")
  
  assign(paste0(i, "_results"), results)
}
```


```{r}
head(M_G1_results)
```

```{r}
decideTests(fit_contrasts, adjust.method = "BH", p.value = 0.01) %>%
  summary()
```


#### Adding user-defined significance thresholds based on FDR and LFC

```{r}

```


* Discussion about p-value correction and why it is needed -> FDR as a method for this
* Discussion about whether a log2FC threshold should also be set -> mention MS2 TMT ratio compression



### Visualising the results of our statistical model

* Volcano plot
* MA plot
* Adding error bars to show confidence interval 

::: {.callout-tip}
#### Key Points

- Last section of the page is a bulleted summary of the key points
:::
