---
title: Biological interpretation
bibliography: course_refs.bib
editor_options: 
  chunk_output_type: console
---

::: {.callout-tip}
#### Learning Objectives

* Be aware of different analyses that can be done to gain a biological understanding of expression proteomics results
* Understand the concept of Gene Ontology (GO) enrichment analyses 
* Complete GO enrichment analyses using the `enrichGO` function from the `clusterProfiler` package

:::

```{r, echo = FALSE, fig.align = "center", out.width = "90%"}
knitr::include_graphics("figs/flow_chart/flow_chart.012.png", error = FALSE)
```


```{r, eval=TRUE, include=FALSE}
library("QFeatures")
library("NormalyzerDE")
library("limma")
library("factoextra")
library("org.Hs.eg.db")
library("clusterProfiler")
library("enrichplot")
library("patchwork")
library("tidyverse")
library("here")

load(here("course_files/preprocessed/lesson06.rda"), verbose = TRUE)
```


## Adding metadata to our results using `dplyr`

Before we can look any further into the biological meaning of any protein 
abundance changes we need to extract these proteins from our overall results.
It is also useful to re-add information about the master protein descriptions
since this is lost in the output of `limma` analysis. 

It is important to note that the results table we have generated from `limma`
is not in the same order as the input data. Therefore, to add information from our
original data e.g., from the `rowData` such as the `Master.Protein.Descriptions`
we must match the protein names between them. 

To do this, let's extract information the `Master.Protein.Descriptions` from the original
data we have created called `all_proteins`. 

Recall that `all_protiens` is `SummarizedExperiment` object,

```{r}
all_proteins
```

We wish to extract information from the `rowData` regarding the
`Master.Protein.Descriptions`,

```{r}
## Add master protein descriptions back
protein_info <- all_proteins %>%
  rowData() %>%
  as_tibble() %>%
  select(Protein = Master.Protein.Accessions, 
         protein_description = Master.Protein.Descriptions)

protein_info %>% head()
```

Note, we also extract the `Master.Protein.Accessions` column so we can use
this to match to the `protein` column in our `limma` results.

Now we can use the `left_join` function from `dplyr` to match the protein
descriptions to the protein IDs,

```{r, message=FALSE, warning=FALSE}
limma_results <- limma_results %>% 
  left_join(protein_info, by = "Protein")

# Verify
limma_results %>%
  head()
```

::: {.callout-note}
#### Manipulating data with `dplyr` and `tidyverse`
There is lots of information online about getting started with `dplyr` and
using the `tidyverse`. We really like [this lesson from the Data Carpentry](https://datacarpentry.org/R-ecology-lesson/03-dplyr.html) if
you are new to the `tidyverse`.

:::

## Subset differentially abundant proteins

Let's subset our results and only keep proteins which have been flagged as 
exhibiting significant abundance changes, 

```{r}
sig_changing <- limma_results %>% 
  as_tibble() %>%
  filter(significance == "sig")

sig_up <- sig_changing %>%
  filter(direction == "up")

sig_down <- sig_changing %>%
  filter(direction == "down")
```


## Biological interpretation of differentially abundant proteins

Our statistical analyses provided us with a list of proteins that are present
with significantly different abundances between M-phase and G1-phase of the cell 
cycle. We can get an initial idea about what these proteins are and do by 
looking at the protein descriptions. 

```{r}
## Look at descriptions of proteins upregulated in M relative to deysnchronized cells
sig_up %>%
  pull(protein_description) %>%
  head()
```

Whilst we may recognise some of the changing proteins, this might be the first
time that we are coming across others. Moreover, some protein descriptions 
contain useful information, but this is very limited. We still want to find out
more about the biological role of the statistically significant proteins so that
we can infer the potential effects of their abundance changes. 

There are many functional analyses that could be done on the proteins with
differential abundance:

* Investigate the biological pathways that the proteins function within ([KEGG](https://www.genome.jp/kegg/) etc.)
* Identify potential interacting partners ([IntAct](https://www.ebi.ac.uk/intact/), [STRING](https://string-db.org] etc.))
* Determine the subcellular localisation in which the changing proteins are found 
* Understand the co-regulation of their mRNAs ([Expression Atlas](https://www.ebi.ac.uk/gxa))
* Compare our changing proteins to those previously identified in other proteomic
studies of the cell cycle

### Gene Ontology (GO) enrichment analysis 

One of the common methods used to probe the biological relevance of proteins
with significant changes in abundance between conditions is to carry out Gene
Ontology (GO) enrichment, or over-representation, analysis.

The Gene Ontology consortium have defined a set of hierarchical descriptions to
be assigned to genes and their resulting proteins. These descriptions are split
into three categories: cellular components (CC), biological processes (BP) and
molecular function (MF). The idea is to provide information about a protein's
subcellular localisation, functionality and which processes it contributes to
within the cell. Hence, the overarching aim of GO enrichment analysis is to
answer the question:

*"Given a list of proteins found to be differentially abundant in my phenotype of interest, what are the cellular components, molecular functions and biological processes involved in this phenotype?".*

Unfortunately, just looking at the GO terms associated with our differentially
abundant proteins is insufficient to draw any solid conclusions. For example, if
we find that 120 of the `r nrow(sig_down)` proteins significantly
downregulated in M phase are annotated with the GO term "kinase activity", it may
seem intuitive to conclude that reducing kinase activity is important for the M-phase 
phenotype. However, if 90% of all proteins in the cell were kinases (an extreme
example), then we might expect to discover a high representation of the "kinase
activity" GO term in any protein list we end up with.

This leads us to the concept of an over-representation analysis. We wish to ask 
whether any GO terms are over-represented (i.e., present at a higher frequency 
than expected by chance) in our lists of differentially abundant proteins. In 
other words, we need to know how many proteins with a GO term *could* have
shown differential abundance in our experiment vs. how many proteins with this
GO term *did* show differential abundance in our experiment. 

We are going to use a function in R called `enrichGO` from the the
[`clusterProfiler`](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
@Yu2012 Bioconductor R package to perform GO enrichment analysis. The package vignette 
can be found
[here](https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html).
and full tutorials for using the package [here](https://yulab-smu.top/biomedical-knowledge-mining-book/index.html)

::: {.callout-note}
#### Annotation packages in Bioconductor
The `enrichGO` function uses the [`org.Hs.eg.db`](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html) Bioconductor package that has
genome wide annotation for human, primarily based on mapping using Entrez Gene
identifiers. It also uses the [`GO.db`](`https://bioconductor.org/packages/release/data/annotation/html/GO.db.html`) package
which is a set of annotation maps describing the entire Gene Ontology assembled using data from GO.

:::

In the next code chunk we call the `enrichGO` function,

```{r} 
ego_down <- enrichGO(gene = sig_down$Protein,               # list of down proteins
                     universe = limma_results$Protein,      # all proteins 
                     OrgDb = org.Hs.eg.db,                  # database to query
                     keyType = "UNIPROT",                   # protein ID encoding
                     pvalueCutoff = 0.05, 
                     qvalueCutoff = 0.05,
                     ont = "ALL",                           # include CC, BP and MF
                     readable = TRUE)
``` 


Let's take a look at the output.

```{r}
ego_down
```

The output of the `enrichGO` function is an object of class `enrichResult` that
contains the `ID` and `Description` of all enriched GO terms. There is also 
information about which `geneID`s from our significantly downregulated proteins 
are annotated with each of the enriched GO terms. Let's take a look at the 
descriptions.

```{r}
ego_down$Description
```

There is a long list because of the hierarchical nature of GO terms. The results
of GO enrichment analysis can be visualised in many different ways. For a full
overview of GO enrichment visualisation tools see [*Visualization of functional enrichment result*](https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html).


```{r}
dotplot(ego_down, 
        x = "Count", 
        showCategory = 15, 
        font.size = 10,
        color = "p.adjust")
```


::: {.callout-tip}
#### Key Points

- Gene ontology (GO) terms described the molecular function (MF), biological processes (BP) and cellular component (CC) of proteins.
- GO terms are hierarchical and generic. They do not relate to specific biological systems e.g., cell type or condition.
- GO enrichment analysis aims to identify GO terms that are present in a list of proteins of interest (foreground) at a higher frequency than expected by chance based on their frequency in a background list of proteins (universe). The universe should be a list of all proteins included identified and quantified in your experiment.
- The `enrichGO` function from the `clusterProfiler` package provides a convenient way to carry out reproducible GO enrichment analysis in R.
:::

## References {-}
