<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>11&nbsp; Statistical analysis of all cell cycle stages</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/10_lfq_vs_tmt.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Analysis of expression proteomics data in R</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/CambridgeCentreForProteomics/course_expression_proteomics" rel="" target=""><i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs" rel="" target=""><i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../materials/08_maxquant_notes.html">Extended materials</a></li><li class="breadcrumb-item"><a href="../materials/11_statistical_analysis_all_stages.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Statistical analysis of all cell cycle stages</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Welcome</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data &amp; Setup</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Expression proteomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The use-case data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/02_import_and_infrastructure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Import and infrastructure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03_data_processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data cleaning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04_normalisation_aggregation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data normalisation and data aggregation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/05_protein_exploration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Exploration and visualisation of protein data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/06_statistical_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Statistical analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/07_biological_interpretation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Biological interpretation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Extended materials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/08_maxquant_notes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Adapting this workflow to MaxQuant processed data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/09_normalyzer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Exploring normalisation methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/10_lfq_vs_tmt.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Adapting this workflow to label-free proteomics data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/11_statistical_analysis_all_stages.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Statistical analysis of all cell cycle stages</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link active" data-scroll-target="#differential-expression-analysis"><span class="header-section-number">11.1</span> Differential expression analysis</a>
  <ul class="collapse">
<li><a href="#modelling-with-or-without-an-intercept" id="toc-modelling-with-or-without-an-intercept" class="nav-link" data-scroll-target="#modelling-with-or-without-an-intercept"><span class="header-section-number">11.1.1</span> Modelling with or without an intercept</a></li>
  <li><a href="#what-are-contrasts" id="toc-what-are-contrasts" class="nav-link" data-scroll-target="#what-are-contrasts"><span class="header-section-number">11.1.2</span> What are contrasts?</a></li>
  <li><a href="#running-an-empirical-bayes-moderated-test-using-limma" id="toc-running-an-empirical-bayes-moderated-test-using-limma" class="nav-link" data-scroll-target="#running-an-empirical-bayes-moderated-test-using-limma"><span class="header-section-number">11.1.3</span> Running an empirical Bayes moderated test using <code>limma</code></a></li>
  <li><a href="#accessing-the-model-results" id="toc-accessing-the-model-results" class="nav-link" data-scroll-target="#accessing-the-model-results"><span class="header-section-number">11.1.4</span> Accessing the model results</a></li>
  <li><a href="#qc-plots-for-statistical-test-assumptions" id="toc-qc-plots-for-statistical-test-assumptions" class="nav-link" data-scroll-target="#qc-plots-for-statistical-test-assumptions"><span class="header-section-number">11.1.5</span> QC plots for statistical test assumptions</a></li>
  </ul>
</li>
  <li>
<a href="#interpreting-the-overall-linear-model-output" id="toc-interpreting-the-overall-linear-model-output" class="nav-link" data-scroll-target="#interpreting-the-overall-linear-model-output"><span class="header-section-number">11.2</span> Interpreting the overall linear model output</a>
  <ul class="collapse">
<li><a href="#interpreting-the-results-of-a-single-contrast" id="toc-interpreting-the-results-of-a-single-contrast" class="nav-link" data-scroll-target="#interpreting-the-results-of-a-single-contrast"><span class="header-section-number">11.2.1</span> Interpreting the results of a single contrast</a></li>
  <li><a href="#visualising-the-results-of-our-single-contrast-test" id="toc-visualising-the-results-of-our-single-contrast-test" class="nav-link" data-scroll-target="#visualising-the-results-of-our-single-contrast-test"><span class="header-section-number">11.2.2</span> Visualising the results of our single contrast test</a></li>
  <li><a href="#interpreting-the-results-of-all-contrasts" id="toc-interpreting-the-results-of-all-contrasts" class="nav-link" data-scroll-target="#interpreting-the-results-of-all-contrasts"><span class="header-section-number">11.2.3</span> Interpreting the results of all contrasts</a></li>
  <li><a href="#visualising-the-protein-abundances-in-a-heatmap" id="toc-visualising-the-protein-abundances-in-a-heatmap" class="nav-link" data-scroll-target="#visualising-the-protein-abundances-in-a-heatmap"><span class="header-section-number">11.2.4</span> Visualising the protein abundances in a heatmap</a></li>
  </ul>
</li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Statistical analysis of all cell cycle stages</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Using the <code>limma</code> package (<span class="citation" data-cites="Smyth2004">Smyth (<a href="#ref-Smyth2004" role="doc-biblioref">2004</a>)</span>), design a statistical model to test for differentially abundant proteins between more than two conditions</li>
</ul>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="figs/flow_chart/flow_chart.010.png" class="img-fluid figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This workflow is an adjunct to the <a href="./06_statistical_analysis.html">Statistical analysis</a> section which demonstrates how to perform single comparison statistical with <code>limma</code> (<span class="citation" data-cites="Smyth2004">Smyth (<a href="#ref-Smyth2004" role="doc-biblioref">2004</a>)</span>). Please first read through that material, since it includes background explanations and clarifications which are not repeated here.</p>
</div>
</div>
<section id="differential-expression-analysis" class="level2" data-number="11.1"><h2 data-number="11.1" class="anchored" data-anchor-id="differential-expression-analysis">
<span class="header-section-number">11.1</span> Differential expression analysis</h2>
<p>Having cleaned our data, aggregated from PSM to protein level, completed a log2 transformation and normalised the data, we are now ready to carry out statistical analysis.</p>
<p>The aim of this section is to answer the question: <em>“Which proteins show a significant change in abundance between the multiple cell cycle stages?”</em>.</p>
<p>Our null hypothesis is: <strong>(H0)</strong> The mean protein abundance is the same across cell cycle stages.</p>
<p>Our alternative hypothesis is: <strong>(H1)</strong> The mean protein abundance differs across cell cycle stages.</p>
<p>We will use <code>limma</code> to perform the statistical tests using an <strong>empirical Bayes-moderated linear model</strong>. Please see the <a href="./06_statistical_analysis.html">Statistical analysis</a> section for more details about <code>limma</code> and its application in quantitative proteomics.</p>
<p>Depending on whether the linear model is used to perform single comparisons or multifactorial comparisons, the test statistic will either be a t-value or an F-value, respectively. Here, we will perform both, starting with assessing the overall effect of cell cycle with an F-test. The F-test does not tell us which groups are different to one another, only that the cell cycle stage does affect protein abundance. Later we will see how to perform all the pairwise comparisons with t-tests.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is an F-value?
</div>
</div>
<div class="callout-body-container callout-body">
<p>An F-value is a parametric statistical value used to compare the mean values of <strong>three or more</strong> groups. The F-value is the ratio of the between-group variation (explained variance) and the within-group variance (unexplained variance). The higher the F-value, the more significant the difference between the groups is.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is a t-value?
</div>
</div>
<div class="callout-body-container callout-body">
<p>A t-value is a parametric statistical value used to compare the mean values of <strong>two</strong> groups. The t-value is the ratio of the difference in means to the standard error of the difference in means. The further away from zero that a t-value lies, the more significant the difference between the groups is.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How is the p-value obtained?
</div>
</div>
<div class="callout-body-container callout-body">
<p>A p-value may be obtained from a t-value/F-value by comparing the value against an t-value/F-distribution with the appropriate degrees of freedom.</p>
<ul>
<li>
<strong>degrees of freedom</strong> = the number of observations minus the number of independent variables in the model</li>
<li>
<strong>p-value</strong> = the probability of achieving the t-value/F-value under the null hypothesis i.e., by chance</li>
</ul>
</div>
</div>
<p>Here, we will start by performing a comparison between three groups (M phase, G1 phase and desynchronised) for each protein and obtain an F-value and p-value for each protein.</p>
<p>First, we extract the suitable protein experiment. We will remove the pre-treatment sample, since this condition is not replicated and so not amenable to statistical testing.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># extract the log-normalised experiment from our QFeatures object</span></span>
<span><span class="va">all_proteins</span> <span class="op">&lt;-</span> <span class="va">cc_qf</span><span class="op">[[</span><span class="st">"log_norm_proteins"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># subset to exclude the pre-treatment sample</span></span>
<span><span class="va">all_proteins</span> <span class="op">&lt;-</span> <span class="va">all_proteins</span><span class="op">[</span>, <span class="va">all_proteins</span><span class="op">$</span><span class="va">condition</span> <span class="op">!=</span> <span class="st">'Pre-treatment'</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="modelling-with-or-without-an-intercept" class="level3" data-number="11.1.1"><h3 data-number="11.1.1" class="anchored" data-anchor-id="modelling-with-or-without-an-intercept">
<span class="header-section-number">11.1.1</span> Modelling with or without an intercept</h3>
<p>When investigating the effect of a single explanatory <strong>categorical</strong> variable with more than two levels, a design matrix can be created using either <code>model.matrix(~variable)</code> or <code>model.matrix(~0 + variable)</code>. The former will create a model that includes an intercept term, whilst the latter will not. If an intercept is included, the first level of the variable (here, M) is considered the ‘baseline’ and modeled by the intercept. The subsequent levels of the variable (here, G1 and Desynch) are modeled by additional terms in the model, which capture the difference between each other level and the baseline (M). This approach makes intuitive sense if one group of samples are control samples to which all other groups should be compared. Although comparisons can be made between any pair of groups when using a model with an intercept, it’s less intuitive than a model without an intercept. In this experiment, none of the groups are a control group to compare to, and we wish to compare between all samples. Thus, we will model without an intercept here.</p>
<p>For further guidance on generating design matrices for covariates or continuous explanatory variables, see <a href="https://bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-and-contrast-matrices">A guide to creating design matrices for gene expression experiments</a>.</p>
<p>Below, we define the model matrix without any intercept term using the <code>model.matrix(~0 + variable)</code> formulation.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Ensure that conditions are stored as levels of a factor </span></span>
<span><span class="co">## Explicitly define level order by cell cycle stage</span></span>
<span><span class="va">all_proteins</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">all_proteins</span><span class="op">$</span><span class="va">condition</span>, </span>
<span>                                 levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"M"</span>, <span class="st">"G1"</span>, <span class="st">"Desynch"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Design a matrix containing all factors that we wish to model</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="va">all_proteins</span><span class="op">$</span><span class="va">condition</span></span>
<span><span class="va">m_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rename design matrix columns to make them easier to refer to</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">m_design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Verify</span></span>
<span><span class="va">m_design</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  M G1 Desynch
1 1  0       0
2 1  0       0
3 1  0       0
4 0  1       0
5 0  1       0
6 0  1       0
7 0  0       1
8 0  0       1
9 0  0       1
attr(,"assign")
[1] 1 1 1
attr(,"contrasts")
attr(,"contrasts")$condition
[1] "contr.treatment"</code></pre>
</div>
</div>
</section><section id="what-are-contrasts" class="level3" data-number="11.1.2"><h3 data-number="11.1.2" class="anchored" data-anchor-id="what-are-contrasts">
<span class="header-section-number">11.1.2</span> What are contrasts?</h3>
<p>We also define <strong>contrasts</strong>. The contrasts represent comparisons are of interest to us, e.g M phase vs Desynchronised. This is important since we are not directly interested in the parameter (mean) estimates for each group but rather the differences in these parameter (mean) estimates <em>between</em> groups. The <code>makeContrasts</code> function is used by passing the name we wish to give each contrast and how this contrast should be calculated using column names from the design matrix. We also pass the <code>levels</code> argument to tell R where the column names come from i.e., which design matrix the contrasts are being applied to.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Specify contrasts of interest</span></span>
<span><span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">makeContrasts</span><span class="op">(</span>G1_M <span class="op">=</span> <span class="va">G1</span> <span class="op">-</span> <span class="va">M</span>, </span>
<span>                           M_Des <span class="op">=</span> <span class="va">M</span> <span class="op">-</span> <span class="va">Desynch</span>, </span>
<span>                           G1_Des <span class="op">=</span> <span class="va">G1</span> <span class="op">-</span> <span class="va">Desynch</span>,</span>
<span>                           levels <span class="op">=</span> <span class="va">m_design</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Verify</span></span>
<span><span class="va">contrasts</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Contrasts
Levels    G1_M M_Des G1_Des
  M         -1     1      0
  G1         1     0      1
  Desynch    0    -1     -1</code></pre>
</div>
</div>
</section><section id="running-an-empirical-bayes-moderated-test-using-limma" class="level3" data-number="11.1.3"><h3 data-number="11.1.3" class="anchored" data-anchor-id="running-an-empirical-bayes-moderated-test-using-limma">
<span class="header-section-number">11.1.3</span> Running an empirical Bayes moderated test using <code>limma</code>
</h3>
<p>After we have specified the design matrix and contrasts we wish to make, the next step is to apply the statistical model.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Fit linear model using the design matrix and desired contrasts</span></span>
<span><span class="va">fit_model</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span>object <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">all_proteins</span><span class="op">)</span>, design <span class="op">=</span> <span class="va">m_design</span><span class="op">)</span></span>
<span><span class="va">fit_contrasts</span> <span class="op">&lt;-</span> <span class="fu">contrasts.fit</span><span class="op">(</span>fit <span class="op">=</span> <span class="va">fit_model</span>, contrasts <span class="op">=</span> <span class="va">contrasts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The initial model has now been applied to each of the proteins in our data. We now update the model using the <code>eBayes</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Update the model using the limma eBayes algorithm</span></span>
<span><span class="va">final_model</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span>fit <span class="op">=</span> <span class="va">fit_contrasts</span>, </span>
<span>                      trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="accessing-the-model-results" class="level3" data-number="11.1.4"><h3 data-number="11.1.4" class="anchored" data-anchor-id="accessing-the-model-results">
<span class="header-section-number">11.1.4</span> Accessing the model results</h3>
<p>To get the results for all of our proteins we use the <code>topTable</code> function with the <code>number = Inf</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Format results</span></span>
<span><span class="va">limma_results_all_contrasts</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span></span>
<span>  fit <span class="op">=</span> <span class="va">final_model</span>,   </span>
<span>  coef <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  adjust.method <span class="op">=</span> <span class="st">"BH"</span>,    <span class="co"># Method for multiple hypothesis testing</span></span>
<span>  number <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span> <span class="op">%&gt;%</span>        <span class="co"># Print results for all proteins</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"Protein"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## Verify</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">limma_results_all_contrasts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Protein      G1_M      M_Des     G1_Des    AveExpr        F      P.Value
1  Q9NQW6 -2.077921  0.8130706 -1.2648507  2.7165663 794.7797 5.148901e-15
2  Q9ULW0 -2.152503  0.7854924 -1.3670105  2.5730063 685.1572 1.425086e-14
3  P49454 -1.991326  1.1442558 -0.8470698  2.5489502 598.3086 3.606132e-14
4  O14965 -2.098029  0.9147903 -1.1832391  0.8426435 551.3458 6.308302e-14
5  Q562F6 -3.205603  1.4313091 -1.7742942 -1.2469471 546.9525 6.662987e-14
6  Q15004 -1.355585 -1.0359248 -2.3915102  0.8346698 542.7099 7.027329e-14
     adj.P.Val
1 1.968425e-11
2 2.724051e-11
3 4.477580e-11
4 4.477580e-11
5 4.477580e-11
6 4.477580e-11</code></pre>
</div>
</div>
</section><section id="qc-plots-for-statistical-test-assumptions" class="level3" data-number="11.1.5"><h3 data-number="11.1.5" class="anchored" data-anchor-id="qc-plots-for-statistical-test-assumptions">
<span class="header-section-number">11.1.5</span> QC plots for statistical test assumptions</h3>
<p>First we examine whether there is the expected relationship between abundance and variance, and that the trend line captures this.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plotSA</span><span class="op">(</span>fit <span class="op">=</span> <span class="va">final_model</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>       xlab <span class="op">=</span> <span class="st">"Average log2 abundance"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="11_statistical_analysis_all_stages_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next, we check that the p-value distribution is as expected.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">limma_results_all_contrasts</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">P.Value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="11_statistical_analysis_all_stages_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section></section><section id="interpreting-the-overall-linear-model-output" class="level2" data-number="11.2"><h2 data-number="11.2" class="anchored" data-anchor-id="interpreting-the-overall-linear-model-output">
<span class="header-section-number">11.2</span> Interpreting the overall linear model output</h2>
<p>Having checked that the model we fitted was appropriate for the data, we can now take a look at the results of our test.</p>
<p>As we saw above, <code>topTable</code> will give us the overall output of our linear model. We previously used this function to generate our <code>limma_results_all_contrasts</code> without specifying any value for the <code>coef</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">limma_results_all_contrasts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Protein      G1_M      M_Des     G1_Des    AveExpr        F      P.Value
1  Q9NQW6 -2.077921  0.8130706 -1.2648507  2.7165663 794.7797 5.148901e-15
2  Q9ULW0 -2.152503  0.7854924 -1.3670105  2.5730063 685.1572 1.425086e-14
3  P49454 -1.991326  1.1442558 -0.8470698  2.5489502 598.3086 3.606132e-14
4  O14965 -2.098029  0.9147903 -1.1832391  0.8426435 551.3458 6.308302e-14
5  Q562F6 -3.205603  1.4313091 -1.7742942 -1.2469471 546.9525 6.662987e-14
6  Q15004 -1.355585 -1.0359248 -2.3915102  0.8346698 542.7099 7.027329e-14
     adj.P.Val
1 1.968425e-11
2 2.724051e-11
3 4.477580e-11
4 4.477580e-11
5 4.477580e-11
6 4.477580e-11</code></pre>
</div>
</div>
<p>Interpreting the output of <code>topTable</code> for a multi-contrast model:</p>
<ul>
<li>
<code>G1_M</code>, <code>M_Des</code> and <code>G1_Des</code> = the estimated log2FC for each model contrast</li>
<li>
<code>AveExpr</code> = the average log abundance of the protein across samples</li>
<li>
<code>F</code> = eBayes moderated F-value. Interpreted in the same way as a normal F-value (see above).</li>
<li>
<code>P.Value</code> = Unadjusted p-value</li>
<li>
<code>adj.P.Val</code> = FDR-adjusted p-value (adjusting across proteins but not multiple contrasts)</li>
</ul>
<section id="interpreting-the-results-of-a-single-contrast" class="level3" data-number="11.2.1"><h3 data-number="11.2.1" class="anchored" data-anchor-id="interpreting-the-results-of-a-single-contrast">
<span class="header-section-number">11.2.1</span> Interpreting the results of a single contrast</h3>
<p>We can look at individual contrasts by passing the contrast name to the <code>coef</code> argument in the <code>topTable</code> function. For example, let’s look at the pairwise comparison between M-phase and desynchronised cells. We use the <code>topTable</code> function to get the results of the <code>"M_Des"</code> contrast. We use the argument <code>confint = TRUE</code> so that the our output reports the 95% confidence interval of the calculated log2FC.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">M_Desynch_results</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span>fit <span class="op">=</span> <span class="va">final_model</span>, </span>
<span>                              coef <span class="op">=</span> <span class="st">"M_Des"</span>, </span>
<span>                              number <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>                              adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                              confint <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"Protein"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Verify</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">M_Desynch_results</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Protein      logFC       CI.L       CI.R   AveExpr         t      P.Value
1  P31350 -2.8479188 -3.0751121 -2.6207254 -1.113925 -26.91319 2.350968e-13
2  O00622  1.3935764  1.2583727  1.5287801  1.822783  22.12969 3.333718e-12
3  P46013  0.7490389  0.6686160  0.8294618  4.944857  19.99669 1.303459e-11
4  P11388  1.0847590  0.9671981  1.2023198  4.047165  19.81088 1.477276e-11
5  P49454  1.1442558  1.0202037  1.2683079  2.548950  19.80397 1.484201e-11
6  P11021  0.6622147  0.5840995  0.7403298  5.655878  18.20106 4.586280e-11
     adj.P.Val        B
1 8.987752e-10 20.71754
2 6.372403e-09 18.32184
3 1.134820e-08 17.03986
4 1.134820e-08 16.92071
5 1.134820e-08 16.91625
6 2.922225e-08 15.83230</code></pre>
</div>
</div>
<p>Interpreting the output of <code>topTable</code> for a single contrast:</p>
<ul>
<li>
<code>logFC</code> = the fold change between the mean log abundance in group A and the mean log abundance in group B</li>
<li>
<code>CI.L</code> = the left limit of the 95% confidence interval for the reported log2FC</li>
<li>
<code>CI.R</code> = the right limit of the 95% confidence interval for the reported log2FC</li>
<li>
<code>AveExpr</code> = the average log abundance of the protein across samples</li>
<li>
<code>t</code> = t-statistic derived from the original statistical test (not a t-test)</li>
<li>
<code>P.Value</code> = Unadjusted p-value</li>
<li>
<code>adj.P.Val</code> = FDR-adjusted p-value (adjusted across proteins but not multiple contrasts)</li>
<li>
<code>B</code> = B-statistic representing the log-odds that a protein is differentially abundant between conditions</li>
</ul>
<p>This time the output of <code>topTable</code> contains a t-statistic rather than an F-value. This is because we only told the function to compare two conditions, so the corresponding t-statistic from our linear test is reported. Importantly, however, the p-value adjustment in this case only accounts for multiple tests across our 3823 proteins, not the three different contrasts/comparisons we used the data for. As a result, we could over-estimate the number of statistically significant proteins within this contrast, although this this effect is only likely to become problematic when we have a larger number of contrasts to account for.</p>
</section><section id="visualising-the-results-of-our-single-contrast-test" class="level3" data-number="11.2.2"><h3 data-number="11.2.2" class="anchored" data-anchor-id="visualising-the-results-of-our-single-contrast-test">
<span class="header-section-number">11.2.2</span> Visualising the results of our single contrast test</h3>
<p>Below, we produce a volcano plot to visualisation the statistical test results for the M vs Desynchronised contrast.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">M_Desynch_results</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>significance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.01</span>, <span class="st">"sig"</span>, <span class="st">"not.sig"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">P.Value</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">significance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, stroke <span class="op">=</span> <span class="fl">0.25</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="11_statistical_analysis_all_stages_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="interpreting-the-results-of-all-contrasts" class="level3" data-number="11.2.3"><h3 data-number="11.2.3" class="anchored" data-anchor-id="interpreting-the-results-of-all-contrasts">
<span class="header-section-number">11.2.3</span> Interpreting the results of all contrasts</h3>
<p>To understand the impact of adjusting for multiple hypothesis testing across our comparisons, we can use the <code>decideTests</code> function. This function provides a matrix of values -1, 0 and +1 to indicate whether a protein is significantly downregulated, unchanged or significantly upregulated in a given contrast. For the function to determine significance we have to provide a p-value adjustment method and threshold, here we use the standard Benjamini-Hochberg procedure for FDR adjustment and set a threshold of <code>adj.P.Value &lt; 0.01</code> for significance.</p>
<p>The <code>decideTests</code> function also takes an argument called <code>method</code>. This argument specifies whether p-value adjustment should account only for multiple hypothesis tests across proteins (<code>"separate"</code>) or across both proteins and contrasts (<code>"global"</code>).</p>
<p>Let’s first look at the results when we apply the <code>"separate"</code> method i.e., consider each contrast separately.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu">decideTests</span><span class="op">(</span>object <span class="op">=</span> <span class="va">final_model</span>,</span>
<span>            adjust.method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>            p.value <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>            method <span class="op">=</span> <span class="st">"separate"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">dt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       G1_M M_Des G1_Des
Down    429   438     86
NotSig 3002  2757   3610
Up      392   628    127</code></pre>
</div>
</div>
<p>From this table we can see the number of significantly changing proteins per contrast. For the <code>M_Des</code> comparison the total number of significantly changing proteins is 1066. This should be the same as the number of proteins with an adjusted p-value &lt; 0.01 in our <code>M_Desynch_results</code> object. Let’s check.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">M_Desynch_results</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1066</code></pre>
</div>
</div>
<p>However, if we use the <code>"global"</code> method for p-value adjustment and therefore adjust for both per protein and per contrast hypotheses we may see slightly fewer significant proteins in our <code>M_Des</code> contrast.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">decideTests</span><span class="op">(</span>object <span class="op">=</span> <span class="va">final_model</span>,</span>
<span>            adjust.method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>            p.value <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>            method <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       G1_M M_Des G1_Des
Down    413   403    110
NotSig 3034  2835   3517
Up      376   585    196</code></pre>
</div>
</div>
<p>Unfortunately, there is no way to specify global p-value adjustment accounting for all contrasts when using <code>topTable</code> to look at a single contrast. Instead, we can merge the results from our globally adjusted significance summary (generated using <code>decideTests</code> with <code>method = "global"</code>) with the results of our overall linear model test (generated using <code>topTable</code> with <code>coef = NULL</code>). We demonstrate how to do this in the code below.</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Determine global significance using decideTests</span></span>
<span><span class="va">global_sig</span> <span class="op">&lt;-</span> <span class="fu">decideTests</span><span class="op">(</span>object <span class="op">=</span> <span class="va">final_model</span>, </span>
<span>                          adjust.method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>                          p.value <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                          method <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"protein"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Change column names to avoid conflict when binding</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">global_sig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"sig_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">global_sig</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add the results of global significance test to overall linear model results</span></span>
<span><span class="va">limma_results_all_contrasts</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">left_join</span><span class="op">(</span><span class="va">limma_results_all_contrasts</span>, </span>
<span>                                  <span class="va">global_sig</span>, </span>
<span>                                  by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Protein"</span> <span class="op">=</span> <span class="st">"sig_protein"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Verify</span></span>
<span><span class="va">limma_results_all_contrasts</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Protein      G1_M      M_Des     G1_Des    AveExpr        F      P.Value
1  Q9NQW6 -2.077921  0.8130706 -1.2648507  2.7165663 794.7797 5.148901e-15
2  Q9ULW0 -2.152503  0.7854924 -1.3670105  2.5730063 685.1572 1.425086e-14
3  P49454 -1.991326  1.1442558 -0.8470698  2.5489502 598.3086 3.606132e-14
4  O14965 -2.098029  0.9147903 -1.1832391  0.8426435 551.3458 6.308302e-14
5  Q562F6 -3.205603  1.4313091 -1.7742942 -1.2469471 546.9525 6.662987e-14
6  Q15004 -1.355585 -1.0359248 -2.3915102  0.8346698 542.7099 7.027329e-14
     adj.P.Val sig_G1_M sig_M_Des sig_G1_Des
1 1.968425e-11       -1         1         -1
2 2.724051e-11       -1         1         -1
3 4.477580e-11       -1         1         -1
4 4.477580e-11       -1         1         -1
5 4.477580e-11       -1         1         -1
6 4.477580e-11       -1        -1         -1</code></pre>
</div>
</div>
<p>We now have three additional column, one per contrast, called <code>sig_G1_M</code>, <code>sig_M_Des</code>, and <code>sig_G1_Des</code>. These columns contain -1, 0 or 1 meaning that the protein is significantly downregulated, non-significant or significantly upregulated in the given contrast.</p>
</section><section id="visualising-the-protein-abundances-in-a-heatmap" class="level3" data-number="11.2.4"><h3 data-number="11.2.4" class="anchored" data-anchor-id="visualising-the-protein-abundances-in-a-heatmap">
<span class="header-section-number">11.2.4</span> Visualising the protein abundances in a heatmap</h3>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Extract accessions of significant proteins</span></span>
<span></span>
<span><span class="co"># Summarise how often each protein passes the significance threshold</span></span>
<span><span class="co"># across the 3 contrasts</span></span>
<span><span class="va">n_sig</span> <span class="op">&lt;-</span> <span class="fu">decideTests</span><span class="op">(</span>object <span class="op">=</span> <span class="va">final_model</span>, </span>
<span>                          adjust.method <span class="op">=</span> <span class="st">"BH"</span>, </span>
<span>                          p.value <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                          method <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span>MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify the proteins significant at least once</span></span>
<span><span class="va">sig_proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">n_sig</span><span class="op">)</span><span class="op">[</span><span class="va">n_sig</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>  </span>
<span></span>
<span><span class="co">## Subset quantitative data corresponding to significant proteins</span></span>
<span><span class="va">quant_data</span> <span class="op">&lt;-</span> <span class="va">cc_qf</span><span class="op">[[</span><span class="st">"log_norm_proteins"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">sig_proteins</span>, <span class="op">]</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">assay</span><span class="op">(</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we use the quantitative data to plot a heatmap using <code>pheatmap</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pheatmap</span><span class="op">(</span>mat <span class="op">=</span> <span class="va">quant_data</span>,</span>
<span>         scale <span class="op">=</span> <span class="st">'row'</span>, <span class="co"># Z-score normalise across the rows (proteins)</span></span>
<span>         show_rownames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Too many proteins to show all their names!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="11_statistical_analysis_all_stages_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Key Points
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>The <code>limma</code> package provides a statistical pipeline for the analysis of differential expression (abundance) experiments and can be used for categorical variables with more than two levels.</li>
<li>When performing multiple contrasts, the p-value adjustment for multiple testing should take this into account. <code>decideTests</code> with <code>method = "global"</code> can be used to this end.</li>
</ul>
</div>
</div>
</section></section><section id="references" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="references">References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Smyth2004" class="csl-entry" role="listitem">
Smyth, Gordon K. 2004. <span>“Linear Models and Empirical Bayes Methods for Assessing Differential Expression in Microarray Experiments.”</span> <em>Statistical Applications in Genetics and Molecular Biology</em> 3 (1): 1–25. <a href="https://doi.org/10.2202/1544-6115.1027">https://doi.org/10.2202/1544-6115.1027</a>.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../materials/10_lfq_vs_tmt.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Adapting this workflow to label-free proteomics data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>